source/common/http/conn_manager_impl.cc

```diff
@@ -1102,8 +1157,19 @@ void ConnectionManagerImpl::ActiveStreamDecoderFilter::removeDownstreamWatermark
   parent_.watermark_callbacks_ = nullptr;
 }
 
-void ConnectionManagerImpl::ActiveStreamEncoderFilter::addEncodedData(Buffer::Instance& data) {
-  return parent_.addEncodedData(*this, data);
+Buffer::WatermarkBufferPtr ConnectionManagerImpl::ActiveStreamEncoderFilter::createBuffer() {
+  auto buffer = new Buffer::WatermarkBuffer(Buffer::InstancePtr{new Buffer::OwnedImpl()},
+                                            [this]() -> void { this->responseDataDrained(); },
```

Comment:
nit: potentially just pass the lambdas into a helper function in base class? Basically same code as the decoder side and has duplicated division stuff which should probably be in one place?

Reply:
I can't think of a way to make it clean the other way.    Most callers of the function which creates buffers don't need the lambdas (only needed once for buffer creation) so passing them around all the time seems lousy.  I could have PURE functions which the base class could call for lambdas but it would make it less obvious which buffer is doing what.

Reply:
Sorry I just meant a helper function in `ActiveStreamFilterBase` or a utility function for buffer create that is also used by the router code. In general, I'm not super picky about code duplication, but whenever it involves math that might get out of sync I get a little concerned.